### PR TITLE
ukify: support pesign as alternative to sbsign 

### PR DESCRIPTION
**Pull request by esposem**

### CODE REVIEW

File: man/ukify.xml [hunk 1]
```diff
@@ -211,6 +211,20 @@
         </para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--secureboot-certificate-dir=<replaceable>PATH</replaceable></option></term>
```
Comment (bluca) [inline man/ukify.xml]: would be good to explicitly mention that these two are required when using pesign, and ignored when using sbsign

**Comment by bluca** [comment]
Please add a test for this in src/ukify/test/test_ukify.py

**Comment by esposem** [comment]
> Please add a test for this in src/ukify/test/test_ukify.py

The problem with testing this is, I need to know a certificate nickname present in MoK. Otherwise it doesn't work. Any idea on how to simulate it?

**Comment by bluca** [comment]
> > Please add a test for this in src/ukify/test/test_ukify.py
> 
> The problem with testing this is, I need to know a certificate nickname present in MoK. Otherwise it doesn't work. Any idea on how to simulate it?

Just the signing, not the booting - can't you create the pesign db on the fly? Check how the current test uses sbsign. I don't think MOK needs to be involved for this.

File: src/ukify/ukify.py [hunk 2]
```diff
@@ -747,6 +781,13 @@ def parse_args(args=None):
     p.add_argument('--secureboot-certificate',
                    dest='sb_cert',
                    help='path to certificate file or engine-specific designation for SB signing')
+    p.add_argument('--secureboot-certificate-dir',
+                   dest='sb_certdir',
+                   default='/etc/pki/pesign',
+                   help='path to nss certificate database directory for PE signing. Default is /etc/pki/pesign')
+    p.add_argument('--secureboot-certificate-nick',
+                   dest='sb_cert_nick',
+                   help='nickname of nss certificate database entry to use for PE signing')
```
Comment (medhefgo) [inline src/ukify/ukify.py]: The help texts should have `[sbsign]`/`[pesign]` added to them to signal that you don't have to pass all of them, or this gets really confusing. Also, please call this `--secureboot-certificate-name`.

Or reuse `--secureboot-certificate` and add an explicit `--secureboot-signtool=` or so.

Comment (bluca) [reply src/ukify/ukify.py]: it's common to have both installed, so a -signtool= option probably makes sense

Comment (esposem) [reply src/ukify/ukify.py]: wouldn't it be better to add --use-pesign that requires --secureboot-certificate and otherwise default to sbsign?

Comment (bluca) [reply src/ukify/ukify.py]: I think an option that takes either sbsign (default) or pesign looks nicer, but it's not too important

Comment (esposem) [reply src/ukify/ukify.py]: added `--signtool`

File: src/ukify/test/test_ukify.py [hunk 3]
```diff
@@ -280,6 +280,46 @@ def test_efi_signing(kernel_initrd, tmpdir):
...
+    ])
+
+    try:
+        ukify.check_inputs(opts)
+    except OSError as e:
+        pytest.skip(str(e))
+
+    ukify.make_uki(opts)
+
+    if shutil.which('pesign'):
```
Comment (bluca) [inline src/ukify/test/test_ukify.py]: this is already checked above, so shouldn't be needed again

**Comment by bluca** [comment]
Any chance the two binary files could be generated on the fly instead of checked in? Binary files and git don't mix well, it's only 60 KB but it's a 60KB that we'll bring on forever in the repo, even if the files are deleted, so the repo size only goes up and never down

**Review by bluca — APPROVED** [review]

File: src/ukify/ukify.py [hunk 4]
```diff
@@ -536,50 +538,88 @@
 
     pe.write(output)
 
+def signer_sign(cmd):
+    print('+', shell_join(cmd))
```
[the remark this answers is not in the document]
Comment (esposem) [reply src/ukify/ukify.py:554]: not a sensitive information logging, as it's just the file name

[the remark this answers is not in the document]
Comment (bluca) [reply src/ukify/ukify.py:554]: yeah already dismissed

**Comment by bluca** [comment]
Needs a rebase

File: src/ukify/ukify.py [hunk 5]
```diff
@@ -536,50 +538,95 @@ def pe_add_sections(uki: UKI, output: str):
...
-        '--key', opts.sb_key,
-        '--cert', opts.sb_cert,
+def sbsign_sign(sbsign_tool, input_f, output_f, opts=None):
+    sign_invocation = [
+            sbsign_tool,
+            '--key', opts.sb_key,
+            '--cert', opts.sb_cert,
+            input_f,
+            '--output', output_f,
+        ]
```
Comment (behrmann) [inline src/ukify/ukify.py]: Could you make the indentation here consistent with `pesign_sign`?

File: src/ukify/ukify.py [hunk 6]
```diff
@@ -536,50 +538,95 @@ def pe_add_sections(uki: UKI, output: str):
...
+        '-n', opts.sb_certdir,
+        '-c', opts.sb_cert_name,
+        '-i', input_f,
+        '-o', output_f,
     ]
+    signer_sign(sign_invocation)
 
-    if opts.signing_engine is not None:
-        sbsign_invocation += ['--engine', opts.signing_engine]
+sbverify = {
```
Comment (behrmann) [inline src/ukify/ukify.py]:
```suggestion
SBVERIFY = {
```
PEP8 recommends capital names for [constants](https://peps.python.org/pep-0008/#constants)

File: src/ukify/ukify.py [hunk 7]
```diff
@@ -536,50 +538,95 @@ def pe_add_sections(uki: UKI, output: str):
...
+    'name': 'sbverify',
+    'option': '--list',
+    'output': 'No signature table present',
+}
 
-    sign_kernel = opts.sign_kernel
-    if sign_kernel is None and opts.linux is not None and opts.sb_key:
-        # figure out if we should sign the kernel
-        sbverify_tool = find_tool('sbverify', opts=opts)
+pesigcheck = {
```
Comment (behrmann) [inline src/ukify/ukify.py]:
```suggestion
PESIGCHECK = {
```

**Comment by esposem** [comment]
* applied various feedbacks received (constant name, indentation)
* rebased
* fixed test_ukify since it was skipping both `test_efi_signing` tests

**Comment by bluca** [comment]
looks good to me, waiting for @keszybz

File: src/ukify/ukify.py [hunk 8]
```diff
@@ -1091,16 +1151,20 @@ def finalize_options(opts):
...
             opts.sb_cert = pathlib.Path(opts.sb_cert)
 
-    if bool(opts.sb_key) ^ bool(opts.sb_cert):
-        raise ValueError('--secureboot-private-key= and --secureboot-certificate= must be specified together')
+    if opts.signtool == 'sbsign':
+        if bool(opts.sb_key) ^ bool(opts.sb_cert):
+            raise ValueError('--secureboot-private-key= and --secureboot-certificate= must be specified together when using --signtool=sbsign')
+    else:
+        if not bool(opts.sb_cert_name):
+            raise ValueError('-certificate-name must be specified when using --signtool=pesign')
```
Comment (keszybz) [inline src/ukify/ukify.py]: One dash is missing here.
--certificate-name=.

File: man/ukify.xml [hunk 9]
```diff
@@ -268,7 +278,25 @@
 
           <listitem><para>A path to a certificate to use for signing of the resulting binary. If the
           <varname>SigningEngine=</varname>/<option>--signing-engine=</option> option is used, this may also
-          be an engine-specific designation.</para></listitem>
+          be an engine-specific designation.This option is required by
```
Comment (keszybz) [inline man/ukify.xml]: Space missing here.

File: man/ukify.xml [hunk 10]
```diff
@@ -268,7 +278,25 @@
...
+          <varname>SecureBootSigningTool=sbsign</varname>/<option>--signtool=sbsign</option>. </para></listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term><varname>SecureBootCertificateDir=<replaceable>SB_PATH</replaceable></varname></term>
+          <term><option>--secureboot-certificate-dir=<replaceable>SB_PATH</replaceable></option></term>
+
+          <listitem><para>A path to a nss certificate database directory to use for signing of the resulting binary.
+          This option is optionally used by <varname>SecureBootSigningTool=pesign</varname>/<option>--signtool=pesign</option>.
```
Comment (keszybz) [inline man/ukify.xml]: "option is optionally used" – rephrase this please. Maybe "Takes effect when SecureBootSigningTool=pesign/--signtool=pesign is used."

File: man/ukify.xml [hunk 11]
```diff
@@ -268,7 +278,25 @@
...
+          Defaults to <literal>/etc/pki/pesign</literal>.</para></listitem>
```
Comment (keszybz) [inline man/ukify.xml]: Please use `<filename>` instead of `<literal>`.

File: src/ukify/ukify.py [hunk 12]
```diff
@@ -550,50 +550,95 @@ def pe_add_sections(uki: UKI, output: str):
...
+    ]
+    if 'flags' in tool:
+        cmd.append(tool['flags'])
+
+    print('+', shell_join(cmd))
+    info = subprocess.check_output(cmd, text=True)
+
+    if tool['output'] in info:
+        return True
+    return False
```
Comment (keszybz) [inline src/ukify/ukify.py]: Maybe just `return tool['output'] in info` ?

File: src/ukify/ukify.py [hunk 13]
```diff
@@ -550,50 +550,95 @@ def pe_add_sections(uki: UKI, output: str):
...
+        sign = pesign_sign
+        verify_tool = PESIGCHECK
+
+    sign_args_present = opts.sb_key or opts.sb_cert_name
 
-        # sbverify has wonderful API
-        if 'No signature table present' in info:
-            sign_kernel = True
+    if sign_tool is None and sign_args_present:
+        raise ValueError('%s, required to sign, is not installed!' % opts.signtool)
```
Comment (keszybz) [inline src/ukify/ukify.py]: No, no, please don't use %.
`f'{opts.signtool}, required for signing, is not installed'`

File: src/ukify/ukify.py [hunk 14]
```diff
@@ -913,18 +952,39 @@ def config_example(self) -> tuple[Optional[str], Optional[str], Optional[str]]:
         help = 'OpenSSL engine to use for signing',
         config_key = 'UKI/SigningEngine',
     ),
+    ConfigItem(
+        '--signtool',
+        choices=('sbsign', 'pesign'),
+        dest='signtool',
+        default='sbsign',
+        help='whether to use sbsign or pesign. Default is sbsign.',
```
Comment (keszybz) [inline src/ukify/ukify.py]: Please match the surrounding style that uses `_=_`.

File: src/ukify/ukify.py [hunk 15]
```diff
@@ -913,18 +952,39 @@ def config_example(self) -> tuple[Optional[str], Optional[str], Optional[str]]:
...
+        dest='sb_certdir',
+        default='/etc/pki/pesign',
+        help='required by --signtool=pesign. Path to nss certificate database directory for PE signing. Default is /etc/pki/pesign',
+        config_key = 'UKI/SecureBootCertificateDir',
+    ),
+    ConfigItem(
+        '--secureboot-certificate-name',
+        dest='sb_cert_name',
+        help='required by --signtool=pesign. pesign needs a certificate nickname of nss certificate database entry to use for PE signing',
+        config_key = 'UKI/SecureBootCertName',
```
Comment (keszybz) [inline src/ukify/ukify.py]: Now we have two options, one with "Cert" and one with "Certificate". Please use "Certificate".

File: src/ukify/ukify.py [hunk 16]
```diff
@@ -1091,16 +1151,20 @@ def finalize_options(opts):
...
+        if bool(opts.sb_key) ^ bool(opts.sb_cert):
+            raise ValueError('--secureboot-private-key= and --secureboot-certificate= must be specified together when using --signtool=sbsign')
+    else:
+        if not bool(opts.sb_cert_name):
+            raise ValueError('-certificate-name must be specified when using --signtool=pesign')
 
-    if opts.sign_kernel and not opts.sb_key:
-        raise ValueError('--sign-kernel requires --secureboot-private-key= and --secureboot-certificate= to be specified')
+    if opts.sign_kernel and not opts.sb_key and not opts.sb_cert_name:
+        raise ValueError('--sign-kernel requires either --secureboot-private-key= and --secureboot-certificate= (for sbsign) or --secureboot-certificate-name (for pesign) to be specified')
```
Comment (keszybz) [inline src/ukify/ukify.py]: `--secureboot-certificate-name=`, because the argument is required.

**Review by keszybz — CHANGES_REQUESTED** [review]
Looks very nice. Thanks for doing this. Some stylistic issues to fix though.

File: src/ukify/ukify.py [hunk 17]
```diff
@@ -353,8 +353,10 @@ def find_tool(name, fallback=None, opts=None):
     if shutil.which(name) is not None:
         return name
 
-    return fallback
+    if fallback is None:
+        print("Tool %s not installed!" % name)
```
Comment (keszybz) [inline src/ukify/ukify.py]: Please, just use an f-string.

File: src/ukify/test/test_ukify.py [hunk 18]
```diff
@@ -219,12 +219,15 @@ def test_config_priority(tmp_path):
         DeviceTree = some/path2
         Splash = some/path3
         Uname = 1.2.3
-        EFIArch=arm
+        EFIArch = arm
```
Comment (keszybz) [inline src/ukify/test/test_ukify.py:222]: Not that it matters greatly, but the style is *inconsistent* here on purpose: we want to test various whitespace arrangements. But in fact the variant without spaces is already used in an example earlier, so I it's fine to add the spaces here.

File: src/ukify/test/test_ukify.py [hunk 19]
```diff
@@ -445,6 +454,48 @@ def test_efi_signing(kernel_initrd, tmpdir):
...
+
+    ukify.make_uki(opts)
+
+    # let's check that sbverify likes the resulting file
+    dump = subprocess.check_output([
+        'pesign', '-S',
+        '-i', output,
+    ], text=True)
+
+    assert f'The signer\'s common name is {author}' in dump
```
Comment (keszybz) [inline src/ukify/test/test_ukify.py]: Why not `f"The signer's common name is {author}"` ?

**Review by keszybz — CHANGES_REQUESTED** [review]

**Comment by esposem** [comment]
Why did the test fail? can anyone re-trigger them without me force pushing again?

**Pull request merged**